utils/gcp/object_storage.cc

```diff
@@ -0,0 +1,820 @@
+    // Construct metadata. Could fmt::format, but this is somewhat safer.
+    rjson::value meta = rjson::empty_object();
+    rjson::add(meta, "name", std::string(bucket));
+    rjson::add(meta, "location", std::string(region.empty() ? "US" : region));
+    rjson::add(meta, "storageClass", std::string(storage_class.empty() ? "STANDARD" : storage_class));
+
+    rjson::value uniformBucketLevelAccess = rjson::empty_object();
+    rjson::add(uniformBucketLevelAccess, "enabled", true);
+    rjson::value iamConfiguration = rjson::empty_object();
+    rjson::add(iamConfiguration, "uniformBucketLevelAccess", std::move(uniformBucketLevelAccess));
+    rjson::add(meta, "iamConfiguration", std::move(iamConfiguration));
+
+    co_await create_bucket(project, std::move(meta));
+}
+
+future<> utils::gcp::storage::client::delete_bucket(std::string_view bucket) {
+    gcp_storage.debug("Delete bucket {}", bucket);
+
+    auto path = fmt::format("/storage/v1/b/{}", bucket);
+
+    auto res = co_await _impl->send_with_retry(path
+        , GCP_OBJECT_SCOPE_FULL_CONTROL
+        , ""s
+        , ""s
+        , httpclient::method_type::DELETE
+    );
+
+    switch (res.result()) {
+    case status_type::ok: // mock server sends wrong code, but seems acceptable
+    case status_type::no_content:
+        co_return; // done and happy
+    default:
+        throw failed_operation(fmt::format("Could not delete bucket {}: {}", bucket, res.result()));
+    }
+}
+
+// See https://cloud.google.com/storage/docs/listing-buckets
+future<std::vector<utils::gcp::storage::object_info>> utils::gcp::storage::client::list_objects(std::string_view bucket, std::string_view prefix) {
+    gcp_storage.debug("List bucket {} (prefix={})", bucket, prefix);
+
+    auto path = fmt::format("/storage/v1/b/{}/o", bucket);
+
+    if (!prefix.empty()) {
+        path += fmt::format("?prefix={}", prefix);
+    }
+
+    std::vector<utils::gcp::storage::object_info> result;
+
+    co_await _impl->send_with_retry(path
+        , GCP_OBJECT_SCOPE_READ_ONLY
+        , ""s
+        , ""s
+        , [&](const seastar::http::reply& rep, seastar::input_stream<char>& in) -> future<> {
+            if (rep._status != status_type::ok) {
+                throw failed_operation(fmt::format("Could not list bucket {}: {}", bucket, rep._status));
+            }
+
+            // ensure these are on our coroutine frame.
+            auto bufs = co_await util::read_entire_stream(in);
+            auto root = rjson::parse(std::move(bufs));
+
+            if (rjson::get<std::string>(root, "kind") != "storage#objects"s) {
+                throw failed_operation("Malformed list object reply");
+            }
+
+            auto items = rjson::find(root, "items");
+            if (!items) {
+                co_return;
+            }
+            if (!items->IsArray()) {
+                throw failed_operation("Malformed list object items");
+            }
+            for (auto& item : items->GetArray()) {
```

Comment:
Are we sure scylladb will no tgo haywire with a huge number of objects in the bucket? Maybe we want pagination here?

Reply:
I think it was @bhalevy mentioning a customer with 600k sstables, now think about the S3 backup folder for such a cluster

Reply:
Flat bucket? Even with listing streaming that is a rather hefty operation...